Add documentation for logging. (#2054)
hashicorp/terraform-website: content/plugin/log/filtering.mdx

---
page_title: 'Home - Plugin Development: Filtering Log Output'
description: >-

>   Information on how to filter the logs that your provider outputs, to help
>   find the information you're looking for.

Reviewer: ```suggestion
  Filter your provider's log outputs to find relevant information more quickly.
```

---

# Filtering Log Output

> A lot of different pieces of software are included in a Terraform run: Terraform itself, the SDK or SDKs that providers are built on, and one or more providers. All of these disparate pieces of software can be writing logs, which means that a lot of different logs are being written, and not all of them are always useful in figuring out why something's happening. To help with that, there are various environment variables used to filter log output.

Reviewer: Can we say something up here about how this page is going to list out the environment variables for every distinct thing that is active in the Terraform run and how you can choose how verbose you want the output to be from each of these things (by setting a level).

Also - can we mention the SDK vs. the framework here? I feel like reading this I'm not sure which one (both?) this page is going to apply to.

I just want to set the page up a bit better so users understand what they're getting and what's going on below.

## Valid Levels

Valid levels you can pass to these environment variables are as follows, in order of most verbose to least verbose:

* `TRACE`
* `DEBUG`
* `INFO`
* `WARN`
* `ERROR`

> * `OFF` (turns off logs for that logger)

Reviewer: Can we do more here to explain what people will get when they set these levels? It's hard right now for me to understand which one I would want to pick. Can we have a brief description with each one about what users are going to get?  

Also - does JSON work with every single one of the things below? Can you set something to both JSON and one of the other levels? Or is it the case that if you set it to JSON you automatically get the most verbose thing?

As a special case, `JSON` will output logs in a machine-readable format, though the format is not guaranteed to be stable between releases.

> ## `TF_LOG`

Reviewer: what if we tried a table here? We could have a heading that says "Logging Environment Variables" and then use a table so people can quickly skim for the ones they want, see what they are applied to, etc. I just think right now it's hard for me to see the big picture and understand what I would want to set and when. 

| Variable | Scope | Description | 
| `TF_LOG` | All Logs | Sets the most verbose log level for Terraform, the SDKs, and all of the providers. It overrides all other logging environment variable settings (e.g., `TF_LOG_PROVIDER`). When reporting bugs to issue trackers, we recommend setting `TF_LOG=TRACE` because (why?). |

## `TF_LOG_SDK_PROTO`

The `TF_LOG_SDK_PROTO` environment variable controls the log output for logs written by [`terraform-plugin-go`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go).

## `TF_LOG_SDK_HELPER_SCHEMA`

The `TF_LOG_SDK_HELPER_SCHEMA` environment variable controls the log output for logs written by the `helper/schema` package in [`terraform-plugin-sdk/v2`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema). Only logs written by `terraform-plugin-log` are affected by this environment variable, meaning many logs in that package will still have their level determined by `TF_LOG` or `TF_LOG_PROVIDER`. We aim to reduce the number of these exceptions over time.

## `TF_LOG_SDK_MUX`

The `TF_LOG_SDK_MUX` environment variable controls the log output for logs written by the [`terraform-plugin-mux`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-mux) module.

## Testing

When testing, the above environment variables will only control the log output for the provider under test. `TF_LOG_CORE` will do nothing.

> Setting `TF_ACC_LOG_PATH` to the path you want log output written to will write all the log output--core, SDKs, and all providers in the run--to the path specified at the `TRACE` level.
> 
> Setting `TF_LOG_PATH_MASK` to a file path with a `%s` in the path will write test output to that path, replacing the `%s` with each test's name and writing only that test's output to that file.

Reviewer: ```suggestion

Use the following environment variables to change where Terraform stores the log outputs: 
- TF_ACC_LOG_PATH`: Set a path to write all log output from Terraform, plugin SDKs, and all providers in the run. Terraform will generate the output at the `TRACE` level of verbosity.

- `TF_LOG_PATH_MASK`: Set to a file path with a `%s` in the path to write test output to that path. Replace the `%s` with each test's name to write only that test's output to the file.
```

## Logging Environment Variables

Logging is off for all subsystems by default. Set one or more of the the logging environment variables below to either [`JSON` format](#log-format) or a [verbosity level](#logging-levels) to write logs for a Terraform run.

- To write all logs from Terraform binary, all providers, and all provider SDKs, set `TF_LOG`.

   -> **Note**: We recommend setting `TF_LOG=TRACE` when you report bugs to issue trackers. 

- To write only a subset of logs (e.g., for a single provider), do not set `TF_LOG` and instead set specific logging environment variables for your use case.

Some environment variables take precedence over others. For example, if you set `TF_LOG=TRACE` and `TF_LOG_PROVIDER_AZUREM=WARN`, all provider logs will be written at the `TRACE` level because `TF_LOG` takes precedence.

| Variable | Logs Written By | Precedence and Usage |
| --------------|-------------|-------------|
| `TF_LOG` | All loggers (Terraform, SDKs, providers) | Overrides all other logging environment variables. |
| `TF_LOG_CORE` | Terraform binary | Does not include providers. |

> | `TF_LOG_PROVIDER` | All providers and provider SDKs used during the run | Overrides logging environment variables for specific providers (e.g., `TF_LOG_PROVIDER_AWS`) and provider SDKs (e.g., `TF_LOG_PROTO`). Only affects logs from Terraform CLI runs and does not affect logs from provider [acceptance tests](#testing). |

Reviewer: Okay @bflad I tried to clarify that caveat that we talked about where this only works during CLI runs and not during acceptance testing. I also added more context to the testing section to try to explain a bit more/remind folks of this again.  Thoughts? We can edit more if need be!

- **`OFF`** - Turns off logging for that logger.
- **`ERROR`** - Least verbose, typically provides more detail about user-facing errors.
- **`WARN`** - Typically includes warnings about deprecations or other environment problems.
- **`INFO`** - Can include details about run environment or configuration.
- **`DEBUG`** - Can include additional details about behavior, especially behaviors that may be non-intuitive.
- **`TRACE`** - Most verbose, and typically includes low-level execution steps.

This example sets the loggers for all providers used in the run to `WARN` verbosity.

```
TF_LOG_PROVIDER=WARN
```

-> **Note:** Prior to Terraform v0.15.0, levels besides `TRACE` may not be
entirely reliable. Refer to [this
PR](https://github.com/hashicorp/terraform/pull/26632) for more details.

## Log Format

By default, logs are written as...
TODO: what's the difference between default and JSON format?

`JSON` will output logs in a machine-readable format. The format is not guaranteed to be stable between releases.
TOD: When we pass JSON, what verbosity level will happen?

> TODO: Should we say a bit about how logs are written that makes them esier to parse? Are they prepended with their subsystem? Let's say we did TRACE verbosity - how can they tell which logs are relevant to them/filter?
> 

Reviewer 2: I think diving into this topic is a whole separate section, e.g. "Parsing Logs with jq" -- I think we can probably save this for inclusion later. We'd also need to talk more about the various SDK logging keys available (`tf_provider_addr`, `tf_rpc`, `tf_proto_version`, etc.) too.

```suggestion
```

## Adding Request Metadata

Use the `With()` method to embed values into every log output line associated with a particular request. This lets you more easily filter the output.

The following example sets the `example_id` key set to `123`. Any subsequent `tflog` calls will now include this value in every log line printed for that request.

```
ctx = tflog.With(ctx, "example_id", 123)
```

TODO: Can we include a snippet of example output in a code block so folks can see what this will look like?

> 

Reviewer: OH - this is something you only do when writing logs. Not when actually getting the log output. Not sure why I didn't get that until right this second.
